presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -111,10 +132,39 @@ public void process(Iterable<QueryExecution> queries)
             totalBytes += bytes;
             if (bytes > queryMemoryLimit) {
                 query.fail(new ExceededMemoryLimitException("Query", DataSize.succinctDataSize(queryMemoryLimit, Unit.BYTE)));
+                queryKilled = true;
             }
         }
         clusterMemoryUsageBytes.set(totalBytes);
 
+        boolean shouldKillQuery = nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) == 1 && isClusterOutOfMemory();
+        boolean lastKilledQueryIsGone = lastKilledQuery == null;
+
+        if (lastKilledQuery != null) {
+            ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+            if (generalPool != null) {
+                lastKilledQueryIsGone = generalPool.getQueryMemoryReservations().containsKey(lastKilledQuery);
+            }
+        }
+
+        if (shouldKillQuery && lastKilledQueryIsGone && !queryKilled && killOnOutOfMemory) {
+            // Kill the biggest query in the general pool
+            QueryExecution biggestQuery = null;
+            long maxMemory = -1;
+            for (QueryExecution query : queries) {
+                long bytesUsed = query.getTotalMemoryReservation();
+                if (bytesUsed > maxMemory && query.getMemoryPool().getId().equals(GENERAL_POOL)) {
+                    biggestQuery = query;
+                    maxMemory = bytesUsed;
+                }
+            }
+            if (biggestQuery != null) {
+                biggestQuery.fail(new PrestoException(CLUSTER_OUT_OF_MEMORY, "The cluster is out of memory, and your query was killed. Please try again in a few minutes."));
```

Comment:
Normally, we don't put that into the exception message, because that's already displayed to the user by the CLI, or whatever other program they're using to submit the query.

Right now there's no config option to adjust the policy, but I think that's a great idea! Feel free to submit a pull request that either makes the policy injectable, or adds more config options